compiler/frontend/src/print.rs

```diff
@@ -0,0 +1,304 @@
+            }
+
+            let mut string = "".to_string();
+            if needs_parentheses {
+                string.push('(');
+            }
+
+            let symbol_fits = max_length.fits(length_needed_for_structure - 1 + symbol.len());
+            if symbol_fits {
+                string.push_str(&symbol);
+            } else {
+                string.push('…');
+            }
+
+            if let Some(value) = value {
+                string.push(' ');
+                if symbol_fits {
+                    let value = print(value, Precedence::High, max_length, visitor)?;
+                    if max_length.fits(length_needed_for_structure - 2 + symbol.len() + value.len())
+                    {
+                        string.push_str(&value);
+                    } else {
+                        string.push('…');
+                    }
+                } else {
+                    string.push('…');
+                }
+            }
+            if needs_parentheses {
+                string.push(')');
+            }
+            string
+        }
+        PrintValue::Text(text) => {
+            // - full text
+            // - `…`
+
+            let text = format!("\"{text}\"");
+            if max_length.fits(text.len()) {
+                text
+            } else {
+                "…".to_string()
+            }
+        }
+        PrintValue::Function => {
+            // - `{ … }`
+            // - `…`
+
+            if max_length.fits(5) { "{ … }" } else { "…" }.to_string()
+        }
+        PrintValue::List(list) => {
+            // - all items: `(Foo, Bar, Baz)`
+            // - some items: `(Foo, Bar, + 2 more)`
+            // - no items shown: `(list of 2 items)`
+            // - `…`
+
+            if !max_length.fits(3) {
+                return Some("…".to_string());
+            }
+
+            if list.is_empty() {
+                return Some("(,)".to_string());
+            }
+
+            let list_len = list.len();
+            let mut items = vec![];
+            let mut total_item_length = 0;
+            for item in list {
+                let item = print(item, Precedence::Low, MaxLength::Unlimited, visitor)?;
+                total_item_length += item.len();
+                items.push(item);
+
+                // surrounding parentheses, items, and for each item comma + space
+                if !max_length.fits(2 + total_item_length + items.len() * 2) {
+                    break;
+                }
+            }
+            if items.len() == list_len && max_length.fits(total_item_length + items.len() * 2) {
+                return Some(format!("({})", items.into_iter().join(", ")));
+            }
+
+            // Not all items fit. Try to remove the back ones, showing "+ X more" instead.
+            while let Some(popped) = items.pop() {
+                total_item_length -= popped.len();
+                let extra_text = format!("+ {} more", list_len - items.len());
```

Comment:
- We could calculate the string length first to see whether it would fit without allocating and formatting the string.
- We could first check whether `+ _ more` fits and only format the number if that's the case

The applies to struct formatting below

Reply:
Opened an issue: https://github.com/candy-lang/candy/issues/603